### PR TITLE
build(appsync): fix jsii bug to prevent toString() override w/ dotnet

### PR DESCRIPTION
**Pull request by BryanPan342**
jsii bug that's preventing packaging

<details>
<summary> bug output </summary>

```shell
#STDOUT> Amazon/CDK/AWS/AppSync/Directive.cs(88,32): error CS0115: 'Directive.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.MonoCDK.Experiment/Amazon.CDK.MonoCDK.Experiment.csproj]
#STDOUT> Amazon/CDK/AWS/AppSync/InputType.cs(66,32): error CS0115: 'InputType.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.MonoCDK.Experiment/Amazon.CDK.MonoCDK.Experiment.csproj]
#STDOUT> Amazon/CDK/AWS/AppSync/InterfaceType.cs(64,32): error CS0115: 'InterfaceType.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.MonoCDK.Experiment/Amazon.CDK.MonoCDK.Experiment.csproj]
#STDOUT> Amazon/CDK/AWS/AppSync/InputType.cs(66,32): error CS0115: 'InputType.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.AWS.AppSync/Amazon.CDK.AWS.AppSync.csproj]
#STDOUT> Amazon/CDK/AWS/AppSync/Directive.cs(88,32): error CS0115: 'Directive.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.AWS.AppSync/Amazon.CDK.AWS.AppSync.csproj]
#STDOUT> Amazon/CDK/AWS/AppSync/InterfaceType.cs(64,32): error CS0115: 'InterfaceType.ToString(AuthorizationType[]?)': no suitable method found to override [/tmp/npm-packqZC1CO/Amazon.CDK.AWS.AppSync/Amazon.CDK.AWS.AppSync.csproj]
```

</details>

**Cause**
C# `Object.toString()` does not have parameters in its function. When IIntermediateType functions inherited the `Object.toString()` function, it would override the `toString()` function with parameters causing an error.

**Fix**
Create `bind` methods called `bindToAuthModes` and `bindToGraphqlApi` to alleviate the need for parameterized `toString()` function. 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appsync/lib/schema-base.ts [hunk 1]
```diff
@@ -135,7 +137,14 @@ export interface IIntermediateType {
...
    */
-  readonly intermediateType?: InterfaceType;
+  readonly intermediateType?: IIntermediateType;
+
+  /**
+   * Bind this Intermediate Type to a GraphQL Api
+   *
+   * @param api The binding GraphQL Api [disable-awslint:ref-via-interface]
+   */
+  bindToGraphqlApi(api: GraphqlApi): IIntermediateType;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/lib/schema-base.ts:147]: our `bind` methods typically take in a scope (even if it's not used), and a struct of `Options` (even if it's empty) as well. It's more future friendly. 

They also return a `xyzConfig` object with the properties that are to be bound. This is a slightly different approach.

File: packages/@aws-cdk/aws-appsync/lib/schema-base.ts [hunk 2]
```diff
@@ -221,22 +226,36 @@ export class Directive {
...
 
   private constructor(statement: string, mode?: AuthorizationType) {
     this.statement = statement;
     this.mode = mode;
   }
 
+  /**
+   * Bind this Interface Type to a GraphQL Api
+   *
+   * @param modes the authorization modes
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/lib/schema-base.ts:244]: nit: needed? not sure this is adding much more than the signature itself

File: packages/@aws-cdk/aws-appsync/lib/schema-intermediate.ts [hunk 3]
```diff
@@ -47,13 +47,27 @@ export class InterfaceType implements IIntermediateType {
...
+  protected modes?: AuthorizationType[];
 
   public constructor(name: string, props: IntermediateTypeProps) {
     this.name = name;
     this.definition = props.definition;
     this.directives = props.directives;
   }
 
+  /**
+   * Bind this Interface Type to a GraphQL Api
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/lib/schema-intermediate.ts:62]: update this comment to include when it's called (see bindToCluster, bindToInstance for examples)

File: packages/@aws-cdk/aws-appsync/lib/schema-base.ts [hunk 4]
```diff
@@ -221,22 +226,36 @@ export class Directive {
...
+   */
+  public bindToGraphqlApi(modes?: AuthorizationType[]): Directive {
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-appsync/lib/schema-base.ts]: is the return type of `Directive` here intentional?

**Review by shivlaks — CHANGES_REQUESTED** [review]
the symptoms are useful, can you also include the root cause as part of the commit body.
The commit body also needs to be updated to reflect the implementation.

**Review by MrArnoldPalmer — APPROVED** [review]

File: packages/@aws-cdk/aws-appsync/lib/schema-base.ts (hunk 4 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-appsync/lib/schema-base.ts]: that's worth a comment

**Pull request merged**